modules/calcite/src/main/java/org/apache/ignite/internal/processors/query/calcite/trait/CorrelationTrait.java

```diff
@@ -92,7 +92,12 @@ public boolean correlated() {
         if (trait == this || this == UNCORRELATED)
             return true;
 
-        return equals(trait);
+        if (!(trait instanceof CorrelationTrait))
+            return false;
+
+        CorrelationTrait other = (CorrelationTrait) trait;
+
+        return other.correlated() && other.correlationIds().containsAll(correlationIds());
```

Comment:
> After this change fromTrait correlated by cor1 will satisfy to toTrait correlated by (cor1,cor2), but this doesn't seem legit.

But what's wrong with this statement? I think it's correct. Correlated trait, as far as I understand, means `depends on variable`, so if some node depends on `cor1` and we have a requirement with a dependency for `cor1` and `cor2`, then this node satisfies this requirement (no extra dependency needed). Currently, there are no tests for this case, but I think it can be reproduced with 2 nested `exists` subqueries.

Reply:
Assume we have slightly modified spool that keeps track of provided correlated variables and restarts itself if latter have changed. This spool will preserve only those correlates that were specified.

Now if there is a tree like Node_1(corr1, corr2) -> ... -> Node_N(corr1, corr2), it's possible to put Spool(corr1) anywhere in between Node_1 and Node_N, but this will actually break correlation by corr2 for Node_1.

The point is if there is a correlated stream, the only possible relations should be only those aware of all correlations of this stream.

Reply:
But such spool should require input correlated by `cor1`, and there are no nodes in the chain that satisfy such condition, so the spool can't be inserted here.

Reply:
> But such spool should require input correlated by cor1

Probably this is the key. Currently a spool just takes an input as is.

Am I right that you suggest to interpret correlation not as "the real correlation of the stream", but "the stream is correlated at most by this corrSet"?

Reply:
> but "the stream is correlated at most by this corrSet"?

Yes. I thought that it works currently such a way (since UNCORRELATED satisfy any correlation and filter passes correlation if this correlation contains a superset of required by filter variables)